src/bootstrap/test.rs

```diff
@@ -584,6 +584,70 @@ impl Step for RustdocJS {
     }
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct DocUI {
+    pub host: Interned<String>,
+    pub target: Interned<String>,
+}
+
+impl Step for DocUI {
+    type Output = ();
+    const DEFAULT: bool = true;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun) -> ShouldRun {
+        run.path("src/test/doc-ui")
+    }
+
+    fn make_run(run: RunConfig) {
+        run.builder.ensure(DocUI {
+            host: run.host,
+            target: run.target,
+        });
+    }
+
+    fn run(self, builder: &Builder) {
+        if let Some(ref nodejs) = builder.config.nodejs {
+            let mut found_puppeteer = false;
+            // Check if puppeteer is installed locally.
+            if let Ok(output) = Command::new("npm")
+                                        .args(&["ls", "puppeteer", "puppeteer-core"])
+                                        .current_dir("src/tools/doc-ui")
+                                        .output() {
+                found_puppeteer = String::from_utf8_lossy(&output.stdout).contains("puppeteer");
+            }
+            if !found_puppeteer {
+                // Check if puppeteer is installed "globally".
+                if let Ok(output) = Command::new("npm")
+                                            .args(&["ls", "-g", "puppeteer", "puppeteer-core"])
+                                            .output() {
+                    found_puppeteer = String::from_utf8_lossy(&output.stdout).contains("puppeteer");
+                }
+            }
+            if !found_puppeteer {
```

Comment:
I think the condition is inverted here, you're running the script only if puppeteer was not found?

Reply:
Yes it is inverted, good find!